### ISSUE TITLE
MarkerPlacer when used in MATLAB places markers incorrectly 

### ISSUE THREAD

**Issue by ANKELA** [opensim-org/opensim-core#888]
After scaling a model with ModelScaler, using MarkerPlacer places the markers very far away from where they should be. The resulting model places markers sometimes over a meter from where they should be. I saw another issue that fixed markers are being unchecked as fixed, could this be the problem? They usually are placed over a meter away from where they should be, which makes me think this might be the cause of the problem. 

`
scaleTool.getMarkerPlacer().setApply(true);
scaleTool.getMarkerPlacer().setTimeRange(TimeArray2);
scaleTool.getMarkerPlacer().setStaticPoseFileName(staticPoseMarkerName);
scaleTool.getMarkerPlacer().setOutputModelFileName([scaledModelName '.osim']);
scaleTool.getMarkerPlacer().setOutputMotionFileName(staticCoordinates);
scaleTool.getMarkerPlacer().setOutputMarkerFileName(movedMarkers);
scaleTool.getMarkerPlacer().setMaxMarkerMovement(-1.00000000);
scaleTool.getMarkerPlacer().processModel(model,path2subject);
`

**Comment by jimmyDunne** [comment]
Are you sure that the markers are attached to bodies of the model and not ground? A common error is to accidently have Model markers expressed in ground.

**Comment by ANKELA** [comment]
They are expressed in the bodies they are attached to. I'm loading a marker file pre-version 30505, so each marker looks like this in the loaded file. 

`				<Marker name="R_AC">
					
					<body>torso</body>
					
					<location> -0.05 0.401 0.18</location>
					
					<fixed>false</fixed>
				</Marker>`
Then in the file produced after scaling it looks like this:
`			<Marker name="R_AC">
				
				<connectors>
					<Connector_PhysicalFrame_ name="reference_frame">
						
						<connectee_name>torso</connectee_name>
					</Connector_PhysicalFrame_>
				</connectors>
				
				<location>1.2964587407732162 2.6863950001034302 0.82377125430478593</location>
			</Marker>`
This is the same for every marker though: `<Connector_PhysicalFrame_ name="reference_frame">`
Is it a conversion issue?

**Comment by ANKELA** [comment]
If I run the exact code with OpenSim 3.3 (adding state to the input of processModel) it works as it should.

**Comment by chrisdembia** [comment]
> This is the same for every marker though:
> 
>  <Connector_PhysicalFrame_ name="reference_frame">

That should be the same for every Marker. Every Marker in 4.0 has a "Connector" named "reference_frame". What's important is that the connectee_name is correct.

**Comment by ANKELA** [comment]
connectee_name is correct for all markers.

**Comment by chrisdembia** [comment]
> fixed markers are being unchecked as fixed

Could you clarify? You mean that a Marker in 3.3 with `<fixed>true</fixed>` will be converted in 4.0 to `<fixed>false</fixed>`?

**Comment by chrisdembia** [comment]
> The resulting model places markers sometimes over a meter from where they should be.

Was this observed in XML (in the `location` tag) or in some visualizer/GUI?

**Comment by aymanhab** [comment]
Surprise, surprise we've seen this problem long time ago (sorry for the bad memory ;)
https://github.com/opensim-org/opensim-core/issues/182

**Comment by ANKELA** [comment]
> You mean that a Marker in 3.3 with <fixed>true</fixed> will be converted in 4.0 to <fixed>false</fixed>?
The resulting markers don't show a fixed option at all as far as I can see. I was just thinking that this might be part of the issue, but again I'm not sure. 

> Was this observed in XML (in the location tag) or in some visualizer/GUI?
This is observed in XML, I don't have the 4.0 GUI built. To visualize I'm using 3.3 with 3.3 models for the coordinate files written with 4.0 in MATLAB. 

The issue is the placement of the marker when using markerPlacer in 4.0. Converting a scaled 3.3 model shows placements that are correct. A model scaled with 4.0 creates marker placements that are drastically different, and won't work with IK.

**Comment by aymanhab** [comment]
"<fixed>" is not a property of Markers anymore.

**Comment by ANKELA** [comment]
I wasn't really relying on the "fixed" property, just thought it might be part of the issue. 

I still am having trouble getting the markers to be placed properly when using 4.0. It works just fine with 3.3, so I'm not sure what is happening in the new version.

**Comment by ANKELA** [comment]
Here are the models that result from 1) scaling with 3.3 and converting to 4.0, and 2) scaling with 4.0. Marker placement is very different. 

[ModelsScaled.zip](https://github.com/opensim-org/opensim-core/files/188371/ModelsScaled.zip)

**Comment by jenhicks** [comment]
Related to #465

**Comment by aymanhab** [comment]
Some observations after restoring the "fixed" flag to Marker:
- Model::updateMarkerSet has major parts of it commented out by earlier changes in 4.0 in particular Marker::updateFromMarker we have to decide what is the new/correct behavior in 4.0
- More recent changes from PR #1491 may change behavior as noted by @chrisdembia 
- Other issues that make it hard to test the fixed flag variations:  (ScaleTool allows only const ref to GenericModelMaker so we can't programmatically change MarkerSet in C++).

**Comment by chrisdembia** [comment]
> Other issues that make it hard to test the fixed flag variations: (ScaleTool allows only const ref to GenericModelMaker so we can't programmatically change MarkerSet in C++).

This was my doing I think. Could we add an `updGenericModelMaker()`?

**Comment by aymanhab** [comment]
Sure, that would be great :+1:

**Comment by jenhicks** [comment]
@ANKELA Could you provide the files and steps to produce the issue? Thanks!

**Comment by ANKELA** [comment]
This is the project working folder. https://drive.google.com/open?id=0B6tHKsx00x9-bUVKbEh3WXNNMTA 

The MATLAB script that generated the issue is modelScaler.m. In the script options in the first cell, you can choose whether OpenSim 3.3 or 4.0 is being used, which just keeps the generated models organized and also changes a couple commands in the script since you call them a little differently depending on the version. You can also see the generated models with either 3.3 or 4.0 in the models folder.  It's pretty clean and should be very self explanatory, but let me know if you have any questions. 

Thanks, 
Andy

**Comment by jenhicks** [comment]
@jimmyDunne Can you test whether the bug is fixed in latest master?

**Comment by jimmyDunne** [comment]
Firstly, I ran the ScaleTool on gait2392 though the 4.0-Matlab scipt interface. The moved marker locations are the same as in the outputreferences dir generated from OpenSim 3. 

Reference R.ASIS
``` 					<location>   0.02342148   0.03628304   0.13084011 </location>```
4.0 R.ASIS
``` 					<location>0.023984452680014479 0.036570924839751662 0.13056692187007787</location> ```

Thanks, @ANKELA, for sending some testing data. I simplified your folder so I could do a streamlined test in 4.0. I didn't find the initial issue (large, meter, differences between generic and scaled/moved marker locations).  The moved locations look very reasonable. My suggestion would be that there is an issue somewhere in your code, but since there are so many lines to check, I won't be able to go through it. Perhaps add lines to the testing script until the issue re-surfaces?

I have included both tests, including test matlab scipts in the attached zip. 

[testscale.zip](https://github.com/opensim-org/opensim-core/files/839145/testscale.zip)
